src/frontend/src/utils/scan_range.rs

```diff
@@ -58,6 +61,45 @@ impl ScanRange {
             range: full_range(),
         }
     }
+
+    pub fn try_compute_vnode(
+        &self,
+        dist_key_indices: &[usize],
+        pk_indices: &[usize],
+    ) -> Option<VirtualNode> {
+        if dist_key_indices.is_empty() {
+            return None;
```

Comment:
If there's no distribution key, we will encode the `DEFAULT_VNODE`. Check `table::Distribution::fallback`.

Reply:
Btw, when will there be no distribution key? 🤔

Reply:
> Btw, when will there be no distribution key? 🤔

If the materialize is a singleton, like TopN.

> If None is returned here, all partitions will be read (only one in this case?）. I think it makes some sense and DEFAULT VNODE doesn't need to be known here

You're right. It's okay here to pass all vnodes to the storage table and it will further prune the vnodes to scan.